File: 05-rest-to-grpc/src/sample.js
var express    = require('express');
var bodyParser = require('body-parser');

var app = express();
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());

var router = express.Router();
router.post('/echo', function(req, res) {
  res.json({ msg: req.body.msg });
});

app.use('/api', router);
app.listen(8080);

File: 16-grpc/sample.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
)

func main() {
	http.HandleFunc("/api/echo", func(res http.ResponseWriter, req *http.Request) {
		buf, err := ioutil.ReadAll(req.Body)
		if err != nil {
			http.Error(res, "invalid body", 500)
			return
		}
		defer req.Body.Close()

		var d struct {
			Msg string `json:"msg"`
		}
		err = json.Unmarshal(buf, &d)
		if err != nil {
			http.Error(res, "invalid body", 500)
			return
		}
		fmt.Fprintf(res, `{ "msg": "%s" }`, d.Msg)
	})
	http.ListenAndServe(":8080", nil)
}
